terragraph_planner/los/cylindrical_los_validator.py

```diff
@@ -149,87 +99,106 @@ def _compute_confidence_by_radius(
+        # Pre-compute some variables used in the distance computation
+        ax, ay, az = site1.utm_x, site1.utm_y, none_throws(site1.altitude)
+        bx = site2.utm_x - ax
+        by = site2.utm_y - ay
+        bz = none_throws(site2.altitude) - az
+        b_len_2d_sq = bx * bx + by * by
+        b_len_2d = math.sqrt(b_len_2d_sq)
+        b_len_3d_sq = b_len_2d_sq + bz * bz
+
+        highest_site_altitude = max(az, none_throws(site2.altitude))
         for grid in obstructions:
-            x, y, z = grid
             if (
                 self._los_confidence_threshold == 1.0
-                and z > highest_site_altitude
+                and grid.z > highest_site_altitude
             ):
                 return 0.0
-            z_on_max_top_view_plane = -(a_over_c * x + b_over_c * y + d_over_c)
-            if z > z_on_max_top_view_plane:
-                distance = los_2d_line.distance(Point(x, y))
-            else:
-                distance = self._distance_between_grid_top_and_center_line(
-                    grid,
-                    site1,
-                    site2,
-                    yx_slope,
-                    zx_slope,
-                    xy_slope,
-                    zy_slope,
-                    sum_slope_sq,
-                )
+
+            distance = self._distance_between_grid_and_center_line(
+                grid, ax, ay, az, bx, by, bz, b_len_2d_sq, b_len_2d, b_len_3d_sq
+            )
+            if distance is None:
+                continue
             if distance < minimal_radius:
                 return 0.0
             minimal_distance = min(minimal_distance, distance)
         return minimal_distance / self._fresnel_radius
 
-    def _distance_between_grid_top_and_center_line(
+    def _distance_between_grid_and_center_line(
         self,
         grid: Point3D,
-        site1: LOSSite,
-        site2: LOSSite,
-        yx_slope: Optional[float],
-        zx_slope: Optional[float],
-        xy_slope: Optional[float],
-        zy_slope: Optional[float],
-        sum_slope_sq: float,
-    ) -> float:
+        ax: float,
+        ay: float,
+        az: float,
+        bx: float,
+        by: float,
+        bz: float,
+        b_len_2d_sq: float,
+        b_len_2d: float,
+        b_len_3d_sq: float,
+    ) -> Optional[float]:
         """
-        Use grid center as representative point. Compute the distance from the grid
-        center to the 3d los center line to check if it's in the LOS zone.
+        Use grid center as representative point. Compute the distance from the vertical
+        line that goes through grid center to the 3d los center line to check if LOS
+        is blocked. If the grid is behind the either site, return None.
+
+        The LOS center line is represented as r = a + p * b, and the vertical
+        line is represented as r = c + q * d, where a, b, c, d are 3-D vectors,
+        and p, q are numbers between (0, 1). Since d is vertical, we use d = (0, 0, -1)
```

Comment:
Shouldn't d = (0, 0, -cz)? Wouldn't that mean that q below also needs to be divided by cz?

b x d = (-by*cz, bx*cz, 0) --> you should specify here that you are doing a cross product (probably should use x instead of * for cross product).

Reply:
> Shouldn't d = (0, 0, -cz)? Wouldn't that mean that q below also needs to be divided by cz (although that won't make a difference for q >= 0 check)?

Yes, it should not make difference for q >= 0 check, but would make the computation easier.

> b x d = (-bycz, bxcz, 0) --> you should specify here that you are doing a cross product (probably should use x instead of * for cross product).

Good point. Will change.